aptos-move/aptos-vm/src/verifier/transaction_arg_validation.rs

```diff
@@ -11,28 +11,83 @@ use crate::{
+};
 use move_vm_runtime::session::LoadedFunctionInstantiation;
-use move_vm_types::loaded_data::runtime_types::Type;
+use move_vm_types::{
+    gas::{GasMeter, UnmeteredGasMeter},
+    loaded_data::runtime_types::Type,
+};
 use once_cell::sync::Lazy;
 use std::{
     collections::BTreeMap,
     io::{Cursor, Read},
 };
 
-// A map which contains the structs allowed as transaction input and the
-// validation function for those, if one was needed (None otherwise).
-// The validation function takes the serialized argument and returns
-// an error if the validation fails.
-type ValidateArg = fn(&[u8]) -> Result<(), VMStatus>;
+pub(crate) struct FunctionId {
+    module_id: ModuleId,
+    func_name: &'static IdentStr,
+}
+
+type ConstructorMap = Lazy<BTreeMap<String, FunctionId>>;
+static OLD_ALLOWED_STRUCTS: ConstructorMap = Lazy::new(|| {
+    [("0x1::string::String", FunctionId {
+        module_id: ModuleId::new(AccountAddress::ONE, Identifier::from(ident_str!("string"))),
+        func_name: ident_str!("utf8"),
+    })]
+    .into_iter()
+    .map(|(s, validator)| (s.to_string(), validator))
+    .collect()
+});
 
-static ALLOWED_STRUCTS: Lazy<BTreeMap<String, Option<ValidateArg>>> = Lazy::new(|| {
-    [("0x1::string::String", Some(check_string as ValidateArg))]
-        .into_iter()
-        .map(|(s, validator)| (s.to_string(), validator))
-        .collect()
+static NEW_ALLOWED_STRUCTS: ConstructorMap = Lazy::new(|| {
+    [
+        ("0x1::string::String", FunctionId {
+            module_id: ModuleId::new(AccountAddress::ONE, Identifier::from(ident_str!("string"))),
+            func_name: ident_str!("utf8"),
+        }),
+        ("0x1::object::Object", FunctionId {
+            module_id: ModuleId::new(AccountAddress::ONE, Identifier::from(ident_str!("object"))),
+            func_name: ident_str!("address_to_object"),
+        }),
+        ("0x1::option::Option", FunctionId {
+            module_id: ModuleId::new(AccountAddress::ONE, Identifier::from(ident_str!("option"))),
+            func_name: ident_str!("from_vec"),
+        }),
+        ("0x1::fixed_point32::FixedPoint32", FunctionId {
+            module_id: ModuleId::new(
+                AccountAddress::ONE,
+                Identifier::from(ident_str!("fixed_point32")),
+            ),
+            func_name: ident_str!("create_from_raw_value"),
+        }),
+        ("0x1::fixed_point64::FixedPoint64", FunctionId {
+            module_id: ModuleId::new(
+                AccountAddress::ONE,
+                Identifier::from(ident_str!("fixed_point64")),
+            ),
+            func_name: ident_str!("create_from_raw_value"),
+        }),
+    ]
+    .into_iter()
+    .map(|(s, validator)| (s.to_string(), validator))
+    .collect()
 });
 
+pub(crate) fn get_allowed_structs(struct_constructors_feature: bool) -> &'static ConstructorMap {
+    if struct_constructors_feature {
+        &NEW_ALLOWED_STRUCTS
+    } else {
+        &OLD_ALLOWED_STRUCTS
+    }
+}
+
```

Comment:
@gerben-stavenga had a suggestion that for future additions, we can think about maintain a list of map on chain to track "allowed input types" which would be more efficient than individual feature flags

Reply:
I also have a similar concern. I feel like we are already rebuilding the wheels here. The construction of the values is basically equivalent to the following script
```
script {
    fun main() {
        let arg1 = constructor_1(..);
        let arg2 = constructor_2(..);
        foo(arg1, arg2)
    }
}
```
I'm fine with us implementing the constructors for the few known types in this way right now since we don't have to worry about performance or gas metering. I would become wary if we want to open this up to user types. IMO in that case, it's better to implement some sort of client-side sugaring that generates these constructor scripts instead of having the VM do it.

Reply:
If these scripts are generated client side, the payload would pretty much go from perfectly readable (entry function call) to perfectly opaque (script bytecode)

Reply:
We will be doing this over attributes. This actually has been discussed already last fall, and there are some older design docs somewhere. Lets not be too concerned about this.